src/middlewared/middlewared/plugins/vm.py

```diff
@@ -1168,83 +1123,147 @@ def random_mac(self):
+                        'Size is required.',
+                    )
+
+        if not raw:
+            verrors.add(
+                'vmcreate_container.devices',
+                'A disk device of type RAW is required.',
             )
 
-    @accepts()
-    async def list_images(self):
-        return CONTAINER_IMAGES
+        container_image = CONTAINER_IMAGES.get(data['type'])
+        if container_image is None:
+            verrors.add('vmcreate_container.type', 'Invalid container type.')
 
-    @accepts(Int('job_id'))
-    async def get_download_status(self, job_id):
-        """ Returns the status of the job, if job does not exists it returns False."""
-        job_pool = await self.middleware.call('core.get_jobs', [('method', '=', 'vm.fetch_image')])
-        for __job in job_pool:
-            if __job['id'] == job_id:
-                return __job
-        return False
+        if verrors:
+            raise verrors
 
-    @accepts(Str('vmOS'))
-    async def image_path(self, vmOS):
-        """Return the prebuilt image path or false in case it is not supported."""
-        vm_os = CONTAINER_IMAGES.get(vmOS, None)
-        if vm_os:
-            image_file = vm_os['GZIPFILE']
+        sharefs = await self.middleware.call('vm.get_sharefs')
+        if not sharefs:
+            await self.middleware.call('vm.activate_sharefs')
             sharefs = await self.middleware.call('vm.get_sharefs')
-            file_path = sharefs + '/iso_files/' + image_file
-            if os.path.exists(file_path):
-                if self.vmutils.check_sha256(file_path, vmOS):
-                    self.logger.debug('===> Checksum OK: {}'.format(file_path))
-                    return file_path
-                else:
-                    self.logger.debug('===> Checksum NOK, removing file: {}'.format(file_path))
-                    os.remove(file_path)
-                    return False
-            else:
-                return False
+
+        await self.middleware.run_in_thread(
+            self.__fetch_and_decompress, container_image, sharefs, dest, str(size), job,
+        )
+        job.set_progress(80, 'Creating Docker VM')
+
+        try:
+            vmdata = data.copy()
+            vmdata['vm_type'] = 'Container Provider'
+            vmdata['bootloader'] = 'GRUB'
+            vmdata.pop('type')
+            vmdata.pop('root_password')
+            vm = await self.middleware.call('vm.create', vmdata)
+        except Exception as e:
+            raise CallError(f'Failed to create VM: {e}')
+        return vm
+
+    def __fetch_and_decompress(self, container_image, sharefs, dest, size, job):
+
+        file_path = os.path.join(sharefs, 'iso_files', container_image['GZIPFILE'])
+        for retry in range(3):
+            try:
+                self.__fetch_image(container_image['URL'], file_path, job)
+                break
+            except Exception:
+                pass
```

Comment:
Should be continue?

Reply:
Now that we're using the variable down there, it's clearer you want it to log each time the failure happened, so it should stay as pass.